clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -311,16 +338,27 @@ private boolean isMuted(TopicPartition tp, long now) {
     }
 
     /**
-     * Re-enqueue the given record batch in the accumulator to retry
+     * Re-enqueue the given record batch in the accumulator to retry if the batch has not reached deliveryTimeoutMs.
+     * Remove the batch from the list of inflight batches.
      */
     public void reenqueue(ProducerBatch batch, long now) {
-        batch.reenqueued(now);
-        Deque<ProducerBatch> deque = getOrCreateDeque(batch.topicPartition);
-        synchronized (deque) {
-            if (transactionManager != null)
-                insertInSequenceOrder(deque, batch);
-            else
-                deque.addFirst(batch);
+        if (!batch.hasReachedDeliveryTimeout(deliveryTimeoutMs, now)) {
+            batch.reenqueued(now);
+            maybeRemoveFromInflightBatches(batch);
```

Comment:
Unless I'm missing something, we only remove from the `inFlightBatches` collection when we reenqueue and when we encounter a delivery timeout. Don't we need to remove on completion or failure as well?

Reply:
Good catch. This would have to be called from `Sender.failBatch` and `Sender.completeBatch` as well, similar to how the `transactionManager.removeInflightBatch` is called in those methods.

Reply:
good catch!  updated the code to call `maybeRemoveFromInflightBatches` from  `Sender.failBatch` and `Sender.completeBatch`.